playground/engine.py

```diff
@@ -0,0 +1,818 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+import copy
+import torch
+import os
+from typing import List, Optional, Dict
+from enum import IntEnum
+from pathlib import Path
+from dataclasses import dataclass
+from kaolin.render.camera import Camera, generate_centered_pixel_coords, generate_pinhole_rays
+from threedgrut.utils.logger import logger
+from threedgrut.model.model import MixtureOfGaussians
+from threedgrut.model.background import BackgroundColor
+from playground.tracer import Tracer
+from playground.utils.mesh_io import load_mesh, load_materials, load_missing_material_info, create_procedural_mesh
+from playground.utils.depth_of_field import DepthOfField
+from playground.utils.video_out import VideoRecorder
+from playground.utils.spp import SPP
+from playground.utils.kaolin_future.transform import ObjectTransform
+from playground.utils.kaolin_future.fisheye import generate_fisheye_rays
+
+
+#################################
+##       --- Common ---        ##
+#################################
+
+@dataclass
+class RayPack:
+    rays_ori: torch.FloatTensor
+    rays_dir: torch.FloatTensor
+    pixel_x: Optional[torch.IntTensor] = None
+    pixel_y: Optional[torch.IntTensor] = None
+    mask: Optional[torch.BoolTensor] = None
+
+    def split(self, size=None) -> List[RayPack]:
+        if size is None:
+            return [self]
+        assert self.rays_ori.ndim == 2 and self.rays_dir.ndim == 2, 'Only 1D ray packs can be split'
+        rays_orig = torch.split(self.rays_ori, size, dim=0)
+        rays_dir = torch.split(self.rays_dir, size, dim=0)
+        return [RayPack(ray_ori, ray_dir) for ray_ori, ray_dir in zip(rays_orig, rays_dir)]
+
+
+@dataclass
+class PBRMaterial:
```

Comment:
:( why can't we use kaolin [PBRMaterial](https://github.com/NVIDIAGameWorks/kaolin/blob/master/kaolin/render/materials.py#L40)?

Reply:
This is the exact information we need to pass to the ray tracer:
1. Some fields are padded and reshaped to what the ray tracer expects (i.e. metallicRoughnessTexture is a single 2D texture, kaolin separates them, some textures padded to 4d).
2. Some fields are missing in kaolin: `emissive_factor`, `emissive_map`, `alpha_cutoff`, alpha_mode`